packages/components/src/navigator/navigator-provider/component.tsx

```diff
@@ -0,0 +1,96 @@
+/**
+ * External dependencies
+ */
+// eslint-disable-next-line no-restricted-imports
+import type { Ref } from 'react';
+
+/**
+ * WordPress dependencies
+ */
+import { useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import {
+	contextConnect,
+	useContextSystem,
+	WordPressComponentProps,
+} from '../../ui/context';
+import { View } from '../../view';
+import { NavigatorContext } from '../context';
+import type { NavigatorProviderProps, NavigatorPath } from '../types';
+
+function NavigatorProvider(
+	props: WordPressComponentProps< NavigatorProviderProps, 'div' >,
+	forwardedRef: Ref< any >
+) {
+	const { initialPath, children, ...otherProps } = useContextSystem(
+		props,
+		'NavigatorProvider'
+	);
+
+	const [ path, setPath ] = useState< NavigatorPath >( {
+		path: initialPath,
+	} );
+
+	return (
+		<View ref={ forwardedRef } { ...otherProps }>
+			<NavigatorContext.Provider value={ [ path, setPath ] }>
+				{ children }
+			</NavigatorContext.Provider>
+		</View>
+	);
+}
+
+/**
+ * The `NavigatorProvider` component allows rendering nested panels or menus (via the `NavigatorScreen` component) and navigate between these different states (via the `useNavigator` hook).
+ * The Global Styles sidebar is an example of this. The `Navigator*` family of components is _not_ opinionated in terms of UI, and can be composed with any UI components to navigate between the nested screens.
+ *
+ * @example
+ * ```jsx
+ * import {
+ *   __experimentalNavigatorProvider as NavigatorProvider,
+ *   __experimentalNavigatorScreen as NavigatorScreen,
+ *   __experimentalUseNavigator as useNavigator,
+ * } from '@wordpress/components';
+ *
+ * function NavigatorButton( {
+ *   path,
+ *   isBack = false,
+ *   ...props
+ * } ) {
+ *   const navigator = useNavigator();
+ *   return (
+ *   	<Button
+ *   	  onClick={ () => navigator.push( path, { isBack } ) }
+ *   	  { ...props }
+ *   	/>
+ *   );
+ * }
+ *
+ * const MyNavigation = () => (
+ *   <NavigatorProvider initialPath="/">
+ *     <NavigatorScreen path="/">
+ *       <p>This is the home screen.</p>
+ *   	   <NavigatorButton isPrimary path="/child">
+ *          Navigate to child screen.
+ *       </NavigatorButton>
+ *     </NavigatorScreen>
+ *
+ *     <NavigatorScreen path="/child">
+ *       <p>This is the child screen.</p>
+ *       <NavigatorButton isPrimary path="/" isBack>
+ *         Go back
+ *       </NavigatorButton>
+ *     </NavigatorScreen>
+ *   </NavigatorProvider>
+ * );
+ * ```
+ */
+const ConnectedNavigatorProvider = contextConnect(
```

Comment:
What is this for?

Reply:
It connects the component to the Context System, so that the `useContextSystem` can be used to read from the context. It also does more, like forwarding refs, adding a display name and a CSS selector (see [here](https://github.com/WordPress/gutenberg/blob/92662feeb220336fb28f8b79869cb251a7db4e27/packages/components/src/ui/context/context-connect.js)).

It is a core functionality for the new generation of components that we've been building.